toolkit/scripts/check_required_kernel_configs.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Script to check for required kernel configs in a kernel config file
+# Usage: python3 check_required_kernel_configs.py --required_configs <path to json of required configs> --config_file <path to config being checked>
+
+import json
+import argparse
+import sys
+from kernel_sources_analysis import get_data_from_config, get_jsondata_from_jsonfile, extract_kernel_dir_name, extract_config_arch, create_map_of_config_values
+
+def check_required_configs_in_configfile(req_config_json_file, kernel, arch, input_config_map):
+    config_json_data = get_jsondata_from_jsonfile(req_config_json_file)
+    if kernel not in config_json_data:
+        print(f"Kernel {kernel} not found in {req_config_json_file}")
+        print(f"Please provide required configs for {kernel} in {req_config_json_file}")
+        return None
+    required_configs_data = config_json_data[kernel]['required-configs']
+
+    #incorrect configs is map: {config: (newValue, expectedValue, comment, PR)}
+    incorrect_configs = {}
+
+    # go through required configs
+    for config_option, req_value in required_configs_data.items():
+        # check for arch
+        if arch not in req_value['arch']:
+            continue
+        # check required configs present in the kernel's config file
+        if config_option in input_config_map:
+            # check if value is correct
+            if input_config_map[config_option] not in req_value['value']:
+                incorrect_configs[config_option] = {
+                    "newValue": input_config_map[config_option],
+                    "expectedValue": req_value['value'],
+                    "comment": f"Unexpected value: {input_config_map[config_option]}",
+                    "PR": req_value['PR']
+                }
+        # check if required configs removed from the kernel's config file
+        # Note that some required configs are required to be missing
+        elif "" not in req_value['value']: 
+                incorrect_configs[config_option] = {
+                    "newValue": "MISSING",
+                    "expectedValue": req_value['value'],
+                    "comment": f"Config not found.",
+                    "PR": req_value['PR']
+                }
```

Comment:
One thing I'd like to double-check: in the config JSON how do we encode configs with values set to empty strings? For example `CONFIG_DEFAULT_HOSTNAME`? Is it going to be `value: ['""', "some other valid value"]` and `value: [""]` is reserved for configs, which may be missing from the config file altogether?

Reply:
This is a really good question. If you want `CONFIG_X="" ` then the required value has to be `"\"\""` because json only recognises double quotes for values ("") and the config file uses double quotes ("").